### PR TITLE
update icepack to support snowtable better

### PR DESCRIPTION
**Pull request by apcraig**
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Update Icepack to support the snowtable better
- [X] Developer(s): 
    apcraig
- [ ] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    Tested with CICE changes, full test suite on cheyenne with intel has all tests pass but are not bit-for-bit with the current Consortium master (expected).  https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#8a90b84895c3a42806db0345a4183e22715ee066.  This code should be bit-for-bit with the base though.
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on CICE or any other models?
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/.)
    - [X] Yes, but more is needed for snow
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [ ] Please provide any additional information or relevant details below:

### CODE REVIEW

File: columnphysics/icepack_parameters.F90 [hunk 1]
```diff
@@ -771,11 +777,18 @@ subroutine icepack_init_parameters(   &
          rhosmax_in, &      ! maximum snow density (kg/m^3)
          windmin_in, &      ! minimum wind speed to compact snow (m/s)
          drhosdwind_in, &   ! wind compaction factor (kg s/m^4)
-         snwlvlfac_in, &    ! fractional increase in snow depth
+         snwlvlfac_in       ! fractional increase in snow depth
+
+      integer (kind=int_kind), intent(in), optional :: &
```
Comment (eclare108213) [inline columnphysics/icepack_parameters.F90:782]: thanks for noticing and fixing this

File: columnphysics/icepack_snow.F90 [hunk 2]
```diff
@@ -51,33 +60,96 @@ subroutine icepack_init_snow
...
+      ! for testing Icepack without netcdf,
+      ! use a subsampled, hard-coded table
+      !   5 temperatures from 243.15 by 5.0 K
+      !   5 temperature gradients from 0 to 40 K/m by 10
+      !   1 snow density at 300 kg/m3
+
+      ! if snw_aging_table = 'file'
+      ! all data has to be passed into icepack_parameters
+
+      ! tcraig, should be 223.15 and 243.15 below
```
Comment (eclare108213) [inline columnphysics/icepack_snow.F90:88]: Ah, this is what's different with when reading the file.  I think these should be fixed here, now.
If we do that, then is there any reason to have the 'snicar' option?  Speed?

File: columnphysics/icepack_snow.F90 [hunk 3]
```diff
@@ -966,43 +1089,56 @@ subroutine snow_dry_metamorph (nslyr,nilyr, dt, rsnw, drsnw_dry, zqsn, &
...
+         if (lin_T) then
+            T_idx    = nint(abs(zTsn(k)+ Tffresh - min_T   ) / del_T   , kind=int_kind) + 1
+         else
+            call icepack_warnings_setabort(.true.,__FILE__,__LINE__)
+            call icepack_warnings_add(subname//'ERROR: nonlinear lookup table for T not supported yet')
+            return
+         endif
+
+         ! boundary check:
+         rhos_idx = min(isnw_rhos, max(1,rhos_idx))
```
Comment (eclare108213) [inline columnphysics/icepack_snow.F90:1127]: These indices are different by +1.  Is this a bug in the original code?

File: doc/source/icepack_index.rst [hunk 4]
```diff
@@ -411,7 +411,13 @@ either Celsius or Kelvin units).
...
    "sil", "silicate concentration", "mmol/m\ :math:`^3`"
    "sk_l", "skeletal layer thickness", "0.03 m"
-   "snoice", "snow-ice formation", "m"
+   "snowage_drdt0", "snowage table 3D data for drdt0 (10^-6 m/hr)", ""
+   "snowage_kappa", "snowage table 3D data for kappa (10^-6 m)", ""
+   "snowage_rhos", "snowage table dimension data for rhos (kg/m^3)", ""
+   "snowage_T", "snowage table dimension data for temperature (deg K)", ""
+   "snowage_tau", "snowage table 3D data for tau (10^-6 m)", ""
+   "snowage_Tgrd", "snowage table dimension data for temp gradient (deg K/m)", ""
+   "snoice"", "snow-ice formation", "m""
```
Comment (eclare108213) [inline doc/source/icepack_index.rst:420]: too many "" in this line, after snoice and after m

**Review by eclare108213 — APPROVED** [review]
Looks great.  Again, a few comments/questions that can be addressed later.  Approving now so I can merge/test.

File: columnphysics/icepack_snow.F90 (hunk 2 above)
Comment (apcraig) [reply columnphysics/icepack_snow.F90:88]: Actually, the documentation isn't quite correct.  The 11 temperatures goe from 223 to 273 by 5 while on the actual snicar file they go from 223.15 to 273.15 by 5.  That is the "bug" in the hardwired "snicar" values.  

The only reason to keep snicar is if there are table input file that do not have the 1D data on them.  Or if we want to preserve the old 1D values which may not be correct.

File: columnphysics/icepack_snow.F90 (hunk 3 above)
Comment (apcraig) [reply columnphysics/icepack_snow.F90:1127]: I moved the +1 into the index calculation.  If you look at the previous implementation, it was part of the index check in the next code block.  I thought it was clearer to put the +1 in the computation of the index then check the index cleanly vs min/max after.

**Comment by eclare108213** [comment]
Should probably also update icepack_in and set_nml.snw* to be consistent with CICE's.  Will make a note to do that...

File: doc/source/icepack_index.rst (hunk 4 above)
Comment (apcraig) [reply doc/source/icepack_index.rst:420]: I can fix these things.  Do you want me to update the PR or do you want to fix them as you move forward?

Comment (eclare108213) [reply doc/source/icepack_index.rst:420]: Please go ahead and fix them.  Thx!

File: columnphysics/icepack_snow.F90 (hunk 2 above)
Comment (eclare108213) [reply columnphysics/icepack_snow.F90:88]: @njeffery
@apcraig turned up an inconsistency between the snow aging file that I gave him (which was the most recent one on LANL IC machines, snicar_drdt_bst_fit_60_c04262019.nc), where the code uses 223.0 as the starting T index but the file has 223.15.  My inclination is to fix this in the code, but I want to make sure that the file I'm using is correct.

Alternatively, we can dispense with the index calculation in the code completely, and just read it from the file.  That's much better from a software perspective.  Thoughts?

Comment (njeffery) [reply columnphysics/icepack_snow.F90:88]: I double checked the snicar file and it's correct. Also, the ELM code uses an integer 223 in this expression, not 223.15.

**Pull request merged**